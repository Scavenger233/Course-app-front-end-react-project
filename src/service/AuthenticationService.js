import axios from 'axios'
const COURSE_API_URL = 'http://localhost:8080'

export const USER_NAME_SESSION_ATTRIBUTE_NAME = 'authenticatedUser'

class AuthenticationService {

    executeBasicAuthenticationService(username, password) {
        return axios.get(`${COURSE_API_URL}/basicauth`,
            { headers: { authorization: this.createBasicAuthToken(username, password)} })
    }

    createBasicAuthToken(username, password) {
        return 'Basic ' + window.btoa(username + ":" + password)
    }

    registerSuccessfulLogin(username, password) {
        sessionStorage.setItem(USER_NAME_SESSION_ATTRIBUTE_NAME, username)
        this.setupAxiosInterceptors(this.createBasicAuthToken(username, password))
    }

    isUserLoggedIn() {
        let user = sessionStorage.getItem(USER_NAME_SESSION_ATTRIBUTE_NAME)
        if (user === null) return false
        return true
    }

    logout() {
        sessionStorage.removeItem(USER_NAME_SESSION_ATTRIBUTE_NAME);
        this.setupAxiosInterceptors("");
    }

    setupAxiosInterceptors(token) {
        axios.interceptors.request.use(
            (config) => this.setAuthorizationHeader(config, token)
        )
    }

    setAuthorizationHeader(config, token) {
        if (this.isUserLoggedIn()) {
            config.headers.authorization = token
        }
        return config
    }

    /*setupAxiosInterceptors(token) {
        axios.interceptors.request.use(
            (config) => {
                if (this.isUserLoggedIn()) {
                    config.headers.authorization = token
                }
                return config
            }
        )
    }*/
}

export default new AuthenticationService()