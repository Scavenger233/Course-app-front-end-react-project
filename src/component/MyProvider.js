import React, { Component } from 'react';
import AuthenticationService from '../service/AuthenticationService';

export const MContext = React.createContext();  //exporting context object
class MyProvider extends Component {

state = {isUserLoggedIn: false}
render() {
        this.state.isUserLoggedIn = AuthenticationService.isUserLoggedIn();
        return (
            <MContext.Provider value={
            {   state: this.state,
                setIsUserLoggedIn: (value) => this.setState({
                    isUserLoggedIn: value })}}>
            {this.props.children}   
            </MContext.Provider>)
    }
}

export default MyProvider