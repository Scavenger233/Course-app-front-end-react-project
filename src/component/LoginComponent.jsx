import React, { Component } from 'react'
import AuthenticationService from '../service/AuthenticationService';
import withRouter from './withRouter';
import { MContext } from './MyProvider';

class LoginComponent extends Component {

    constructor(props) {
        super(props)

        this.state = {
            username: 'bytecaptain',
            password: '',
            hasLoginFailed: false,
            showSuccessMessage: false
        }

        this.handleChange = this.handleChange.bind(this)
        this.loginClicked = this.loginClicked.bind(this)
    }

    handleChange(event) {
        this.setState(
            {
                [event.target.name]
                    : event.target.value
            }
        )
    }

    loginClicked(context) {
        AuthenticationService
            .executeBasicAuthenticationService(this.state.username, this.state.password)
            .then(() => {
                AuthenticationService.registerSuccessfulLogin(this.state.username, this.state.password)
                context.setIsUserLoggedIn(true);
                this.props.navigation(`/courses`)
            }).catch(() => {
                context.setIsUserLoggedIn(false);
                this.setState({ showSuccessMessage: false })
                this.setState({ hasLoginFailed: true })
            })
    }

    render() {
        return (
            <div>
                <h1 data-testid="loginHeader">Login</h1>
                <div className="container">
                    {this.state.hasLoginFailed && <div className="alert alert-warning">Invalid Credentials</div>}
                    {this.state.showSuccessMessage && <div>Login Sucessful</div>}
                    User Name: <input data-testid="username" type="text" name="username" value={this.state.username} onChange={this.handleChange} />
                    Password: <input data-testid="password" type="password" name="password" value={this.state.password} onChange={this.handleChange} />
                    <MContext.Consumer>
                        {(context) => (
                            <button data-testid="login" className="btn btn-success" onClick={()=>{this.loginClicked(context)}}>Login</button>
                        )}</MContext.Consumer>
                </div>
            </div>
        )
    }
}

export default withRouter(LoginComponent);