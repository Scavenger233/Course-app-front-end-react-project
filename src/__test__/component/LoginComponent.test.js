import React from 'react';
import { BrowserRouter as Router} from 'react-router-dom';
import {render, fireEvent, waitFor, screen} from '@testing-library/react'
import axios from 'axios';
import MyProvider from '../.././component/MyProvider';

import AuthenticationService from '../../service/AuthenticationService';

jest.mock('axios');

import LoginComponent from '../.././component/LoginComponent';

import renderer from 'react-test-renderer'

const basicToken = "Basic Ynl0ZWNhcHRhaW46Ynl0ZWNhcHRhaW4=";

test("renders correctly", () => {
    const tree = renderer.create(<Router><LoginComponent /></Router>).toJSON();
    expect(tree).toMatchSnapshot();
  });

test('Login with right credential', async() => {
    //TODO update
    let response = {data : {token: basicToken}};
    axios.get.mockResolvedValue(response);

    render(<MyProvider><Router><LoginComponent /></Router></MyProvider>);

    const usernameField = screen.getByTestId('username');

    fireEvent.change(usernameField, { target: { value: 'bytecaptain' }});

    const passwordField = screen.getByTestId('password');

    fireEvent.change(passwordField, { target: { value: 'bytecaptain' }});

    const button = screen.getByTestId("login");

    await waitFor(() => {
        fireEvent.click(button);
    });    

  });

  test('Login with worng credential', async() => {
    //TODO update
    let response = null;
    axios.get.mockResolvedValue(response);

    render(<MyProvider><Router><LoginComponent /></Router></MyProvider>);

    const usernameField = screen.getByTestId('username');

    fireEvent.change(usernameField, { target: { value: 'bytecaptain' }});

    const passwordField = screen.getByTestId('password');

    fireEvent.change(passwordField, { target: { value: 'wrongpassword' }});

    const button = screen.getByTestId("login");

    AuthenticationService.executeBasicAuthenticationService = jest.fn().mockRejectedValue(new Error('Login failed'));

    await waitFor(() => {
        fireEvent.click(button);
    });    

  });