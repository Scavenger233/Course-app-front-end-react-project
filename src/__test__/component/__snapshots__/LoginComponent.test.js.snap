// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <h1
    data-testid="loginHeader"
  >
    Login
  </h1>
  <div
    className="container"
  >
    User Name: 
    <input
      data-testid="username"
      name="username"
      onChange={[Function]}
      type="text"
      value="bytecaptain"
    />
    Password: 
    <input
      data-testid="password"
      name="password"
      onChange={[Function]}
      type="password"
      value=""
    />
    <button
      className="btn btn-success"
      data-testid="login"
      onClick={[Function]}
    >
      Login
    </button>
  </div>
</div>
`;
